consensus/istanbul/celo-ibft-spec.md

```diff
@@ -0,0 +1,486 @@
+upon: <Preprepare, Hc, Rd, V, RCC> from proposer(Hc, Rd) && Sc = AcceptRequest
+  if (Rd > 0 && validRCC(V, RCC)) || (Rd = 0 && RCC = nil)  {
+    Rc ← Rd
+    Vc ← V
+    Sc ← Preprepared
+    bc(<Prepare, Hc, Rd, Vc>)
+  }
+
+// When a participant sees at least 2f+1 prepare or commit messages for a value
+// it will send a commit message for that value.
+upon: M ← { <T, Hc, Rd, Vc> : T ∈ {Prepare, Commit} } && |M| >= 2f+1 && Sc ∈ {AcceptRequest, Preprepared} 
+  Sc ← Prepared
+  PCc ← <PreparedCertificate, M, Vc>
+  bc(<Commit, Hc, Rd, Vc>)
+
+// When a participant sees at least 2f+1 commit messages for a vlaue, they
+// consider that value committed (agreed) and pass the value to the application,
+// which will in turn issue a final committed event if the value is considered
+// valid by the application.
+upon: M ← { <Commit, Hc, Rd, Vc> } && |M| >= 2f+1 && Sc ∈ {AcceptRequest, Preprepared, Prepared} 
+  Sc ← Committed
+  deliverValue(Vc)
+
+// Upon receipt of a round change if that round change is old, send the participant's round
+// change back to the sender to help them catch up. Othewise if there are at least 2f+1 round
+// change messages sharing the same round then switch to it. Otherwise if there are at least
+// f+1 round change messages switch to the higest round that is less than or equal to the top
+// f+1 rounds.
+upon: m<RoundChange, Hc , R, PC> && (PC = nil || validPC(PC)) 
+  if R < Rd {
+    send(<RoundChange, Hc, Rd, PCc>, sender(m))
+  } else if quorumRound() > Rd {
+    Rd ← quorumRound()
+    Rc ← quorumRound()
+    Sc ← AcceptRequest
+    schedule onRoundChangeTimeout(Hc, Rd) after roundChangeTimeout(Rd)
+    if Vc != nil && isProposer(Hc, Rc) {
+      bc(<Preprepare, Hc, Rc, Vc, PCc>)
+    }
+  } else if f1Round() > Rd {
+    Rd ← f1Round() 
+    Sc ← WaitingForNewRound
+    schedule onRoundChangeTimeout(Hc, Rd) after roundChangeTimeout(Rd)
+    bc(<RoundChange, Hc, Rd, PCc>)
+  }
+
+// Functions that modify instance state.
+
+// As long as the round and height have not changed since it was scheduled
+// onRoundChangeTimeout sets the desired round to be one greater than the Current
+// round, sets the current state to be WaitingForNewRound and broadcasts a round
+// change message.
+// 
+// Note: This function is referred to in the code as
+// `handleTimeoutAndMoveToNextRound`, which is misleading because it does not move
+// to the next round, it only updates the desired round and sends a round change
+// message. Hence why it has been renamed here to avoid confusion.
+onRoundChangeTimeout(H, R) {
+  if H = Hc && R = Rc {
+    Rd ← Rc+1
+    Sc ← WaitingForNewRound
+    schedule onRoundChangeTimeout(Hc, Rd) after roundChangeTimeout(Rd)
+    bc<RoundChange, Hc, Rd, PCc>
+  }
+}
+```
+
+### Supporting Functions
+
+These functions can read global state but cannot modify it.
+
+#### Application provided functions
+No pseudocode is provided for these functions since their implementation is
```

Comment:
This is true for many, but for example, what `proposer(H, R)` should return is actually part of the protocol, isn't it?

Reply:
I don't think it is, we actually have 3 different implementations currently, sticky, round robin, and shuffled round robin where the proposer at any point is defined by the implementation of the randomness contract.

Reply:
Agree. The protocol assumes there is a proposer and that it can depend on the heigh and on the round, but it doesn't specify the semantics of the selection, that's outside of the protocol